crates/wasi-common/src/sys/windows/hostcalls_impl/misc.rs

```diff
@@ -77,12 +77,158 @@ pub(crate) fn clock_time_get(clock_id: wasi::__wasi_clockid_t) -> Result<wasi::_
+                // On Unix, ioctl(FIONREAD) will return 0 for stdout/stderr. Emulate the same behavior on Windows.
+                //
+                // Besides, the spec is unclear what nbytes should actually be for __WASI_EVENTTYPE_FD_WRITE and
+                // the implementation on Unix just returns 0 here, so it's probably fine to do the same on Windows for now.
+                // cf. https://github.com/WebAssembly/WASI/issues/148
+                _ => 0,
+            };
+
+            events.push(wasi::__wasi_event_t {
+                userdata: event.userdata,
+                r#type: event.r#type,
+                error: wasi::__WASI_ESUCCESS,
+                u: wasi::__wasi_event_u {
+                    fd_readwrite: wasi::__wasi_event_fd_readwrite_t {
+                        nbytes: size,
+                        flags: 0,
+                    },
+                },
+            })
+        }
+    } else {
+        assert_ne!(stdin_events.len(), 0, "stdin_events should not be empty");
+
+        // We are busy-polling the stdin with delay, unfortunately.
+        //
+        // We'd like to do the following:
+        // (1) wait in a non-blocking way for data to be available in stdin, with timeout
+        // (2) find out, how many bytes are there available to be read.
+        // For one, using `WaitForSingleObject` on the stdin handle could possibly be one way of
+        // achieving (1).
+        // I don't know of any way to achieve (2).
+        //
+        // While both of these are not as trivial on Windows as they are on Linux, there's a much
+        // more fundamental issue preventing us from achieving such behavior with the current
+        // implementation of wasi-common.
+        //
+        // Precisely, in `fd_read` we are using `io::stdin` via the `BufRead` trait, which does
+        // buffering on the libstd side. This means that even if there's still some unread data
+        // in stdin, the Windows system calls may return false negatives, indicating that stdin is empty.
+        // Therefore, avoiding the busy-poll here would require us to ditch libstd for the interaction
+        // with stdin altogether.
+        //
+        // However, polling stdin is a relatively infrequent use case, so this hopefully won't be
+        // a major issue.
+        let poll_interval = Duration::from_millis(10);
+        let poll_start = Instant::now();
+        let timeout_duration = timeout
+            .map(|t| t.delay.try_into().map(Duration::from_nanos))
+            .transpose()?;
+        let timeout = loop {
+            if let Some(timeout_duration) = timeout_duration {
+                if poll_start.elapsed() >= timeout_duration {
+                    break timeout;
+                }
+            }
+            if stdin_nonempty() {
+                break None;
+            }
+            std::thread::sleep(poll_interval);
+        };
+
+        // TODO try refactoring pushing to events
```

Comment:
Agreed, it'd probably be cleaner to have a single loop over `stdin_events` and then set `userdata`, `type`, and `nbytes` depending on timeout or not so that there's a single push to `events`.